### ISSUE TITLE
Empty passwords

### ISSUE THREAD

**Issue by outime** [saltstack-formulas/users-formula#97]
Is there a way to set empty passwords for users? I couldn't see anything like that in the pillar.

**Comment by gravyboat** [comment]
the user.present content (https://github.com/saltstack-formulas/users-formula/blob/master/users/init.sls#L81) supports the syntax that would allow you to omit the password field in the pillar if you so desire and no password will be set for the user.

**Comment by outime** [comment]
Thanks. Would it be a good idea to modify it so it's like...

```
{% if 'password' in user -%}
- password: '{{ user['password'] }}'
{% else -%}
- empty_password: True
{% endif -%}
```

...? I don't know if Salt acts in a different way when the ``empty_password`` is set (can't test it at this very moment).

**Comment by gravyboat** [comment]
Hmm, I think I'd rather add a secondary item in the pillar, so you could do:

```
empty_password: True
```
in your pillar file, then the associated code would look like this:

```
{% if 'password in user -%}
- password: '{{ user['password'] }}'
{% endif -%}
{% if empty_password in user -%}
- empty_password: {{ user['empty_password'] }}
{% endif -%}
```

Then we cover both your user case, as well as ensuring the existing functionality is not modified. Thoughts?

**Comment by outime** [comment]
At first I liked the second example more but then I thought that perhaps you make a typo while writing a pillar and you

**Comment by outime** [comment]
At first I liked the second example more but then I thought that perhaps you make a typo while writing the pillar and you type "Fales" instead of ``False`` and the empty password would be set to ``True``, which could be a security issue. However if we choose the first example it'd result in a parse error which is great since the user would spot the typo. Is this correct?

**Comment by gravyboat** [comment]
So in my 'ideal' scenario, if the user wants to use the setting of `False` they would simply not set that value in the pillar as by default `empty_password` is already a false value and thus setting it to `False` in the pillar is redundant. The first example would indeed result in a parse error.

**Comment by outime** [comment]
I understand what you mean, however I think we should avoid all possible issues. A non-experimented user may think that he needs to specify it because he saw it in the example and he might make that typo. I know this is unlikely to happen... but you know small details like this can stay undetected for a long time. So I keep thinking the first option is better.

**Comment by gravyboat** [comment]
Yeah, we'll let this sit for a day or two and see if anyone else weighs in strongly one way or the other. Thanks for your input!

**Comment by skandyla** [comment]
As for me, empty_password is quite dangerous variable. And makes a big security hole.
```empty_password
    Set to True to enable password-less login for user, Default is False```
So, if user wants to add one, that this should be explicitly defined. Also, personally I use ssh keys options, and even don't touch password settings. (Pam login is managed by ldap). Setting something like this
```
{% if 'password' in user -%}
- password: '{{ user['password'] }}'
{% else -%}
- empty_password: True
{% endif -%}
``` is very, very unwanted scenario.

On my view, user should just add
```empty_password: True``` in case he wants this feature, and omit this setting in other case. I suppose this can be explained in pillar.example, and ```WARNING``` word is strongly recommended here.

By the way, I would rather vote to skip this feature than to add it.

**Comment by 0xf10e** [comment]
IF we add this feature is added we should have it strongly guarded against accidental use:
```
{% if 'password' in user -%} 
    - password: '{{ user['password'] }}' 
{% elif 'empty_password' in user 
  and user['empty_password'] is boolean 
  and user['empty_password'] -%} 
    - empty_password: True 
{% endif -%}
```
  - If a password is set, don't even look at `empty_password`
  - If NO password is set AND
   `empty_password` is a bool 
    evaluating to True then set
    `empty_password`

A warning when a user with an empty password is created would still be nice.

**Comment by outime** [comment]
@skandyla the snippet you're showing is discarded already, the discussion is between these two (or a new better one):

**Option 1**
```
{% if 'password in user -%}
- password: '{{ user['password'] }}'
{% endif -%}
{% if empty_password in user -%}
- empty_password: {{ user['empty_password'] }}
{% endif -%}
```

**Option 2**
```
{% if 'password in user -%}
- password: '{{ user['password'] }}'
{% endif -%}
{% if empty_password in user -%}
- empty_password: True
{% endif -%}
```

On the other hand, I don't see why this feature shouldn't be implemented. It's a SaltStack feature and a formula should just implement all the available SaltStack features.

@0xf10e correct me if I'm wrong but I don't see why we should check for `boolean` since SaltStack will raise a parse error exception if no boolean is set because `empty_password` only accepts booleans itself, so no need to check it in the formula IMHO.

On the other hand, I don't like the idea of ignoring fields when SaltStack doesn't do it (does it?). If ``empty_password`` is ``True``, then it will be ``True``. SaltStack will act according to that value and doesn't check if ``pasword`` is set. We shouldn't implement a different logic from the SaltStack's one or the resulting states will be unpredictable. I hope you can understand what I'm saying here.

And overall I agree with the additional warning.

**Comment by 0xf10e** [comment]
@outime Option 2 in this form it a **BAD** idea, as **any** value set will cause ``empty_password`` to be set to ``True``. Even when the actual value would evaluate to False like ``False``, ``None``, ``''``, ``[]``, ``{}`` do.
I think this line evaluates to ``empty_password = None`` in YAML:
```
 - empty_password:
```
Option 2 would set ``empty_password`` to ``True`` in this case.


You're right about those checks not being necessary when Salt does them.
Has ``states.user.present()`` a defined behaviour for both a password being  specified password and ``empty_password`` being set to ``True`` at the same time?
The [latest docs](https://docs.saltstack.com/en/latest/ref/states/all/salt.states.user.html) don't mention this case.

**Comment by outime** [comment]
@0xf10e both options are a bad idea in that sense, you're right in that I missed the "in user" part but just remove that and we're set. So, something like:

``
{% if 'password in user -%}
- password: '{{ user['password'] }}'
{% endif -%}
{% if empty_password is defined -%}
- empty_password: {{ user['empty_password'] }}
{% endif -%}
``

Would do the trick, also covering possible typos.

About your last question, I don't know either but I don't think we should discuss that - whatever SaltStack does when ``empty_password`` and ``password`` are set is none of our business (from the formula point of view).

**Comment by outime** [comment]
@0xf10e you're right in that I overlooked the "in user" part but just remove that and we're set. So, something like:

```
{% if 'password in user -%}
- password: '{{ user['password'] }}'
{% endif -%}
{% if empty_password is defined -%}
- empty_password: {{ user['empty_password'] }}
{% endif -%}
```

Would do the trick, also covering possible typos.

About your last question, I don't know either but I don't think we should discuss that - whatever SaltStack does when ``empty_password`` and ``password`` are set is none of our business (from the formula point of view).

**Comment by 0xf10e** [comment]
@outime I think this should be OK with ``empty_password`` being set to the provided value.
I also agree, the warning for both parameters being set should happen in ``states.user.present()``.

@gravyboat @skandyla you're OK with this?

(btw: Looking at [line 399](https://github.com/saltstack/salt/blob/develop/salt/states/user.py#L399) and [line 447](https://github.com/saltstack/salt/blob/develop/salt/states/user.py#L447) Salt deletes a set password and also doesn't set one if ``empty_password`` is true.)

**Comment by skandyla** [comment]
Personally, I think this is a good approach. Since it doesn't affect to existing configurations, but allows extra feature.

**Comment by puneetk** [comment]
Is Resolved in #98

**Comment by gravyboat** [comment]
Solution works fine for me, thanks for the input everyone.
